### PR TITLE
MacOS frida ASAN fix

### PR DESCRIPTION
**Pull request by mineo333**
This patch makes startup on MacOS significantly faster as well as fixes some bugs. The way we increase the speed by stopping at the shadow region as well as it is the highest possible relevant region. 

We also remove TLS tracking on darwin systems as TLS regions are allocated on the heap.

### CODE REVIEW

**Comment by s1341** [comment]
You did a lot of other changes, in addition to the memory enumeration changes.

Can you please redo the PR as *only* the memory enumeration changes?

**Comment by mineo333** [comment]
@s1341 I've tried to remove most of the unnecessary changes, but I've found that these are necessary as otherwise either the tests won't build/pass or the example fuzzers won't build.

File: libafl_frida/src/asan/asan_rt.rs [hunk 1]
```diff
@@ -421,7 +421,7 @@ impl AsanRuntime {
 
         let range = Self::range_for_address(stack_address);
 
-        assert!(!(range.0 == 0), "Couldn't find stack mapping!");
+        assert!(range.0 != 0, "Couldn't find stack mapping!");
```
Comment (domenukk) [inline libafl_frida/src/asan/asan_rt.rs:424]: Could do `assert_ne!(range.0, 0)`

File: libafl_frida/src/alloc.rs [hunk 2]
```diff
@@ -453,6 +453,7 @@ impl Allocator {
         let shadow_addr = map_to_shadow!(self, (address as usize));
         let shadow_size = size >> 3;
         let buf = unsafe { std::slice::from_raw_parts_mut(shadow_addr as *mut u8, shadow_size) };
+        //log::trace!("aligned shadow buf: {:?}", buf);
```
Comment (s1341) [inline libafl_frida/src/alloc.rs:456]: Can we get rid of these commented `log::trace`s, everywhere.

File: libafl_frida/src/alloc.rs [hunk 3]
```diff
@@ -568,12 +569,12 @@ impl Allocator {
...
                 let start = range.memory_range().base_address().0 as usize;
                 let end = start + range.memory_range().size();
-                if self.is_managed(start as *mut c_void) {
-                    log::trace!("Not unpoisoning: {:#x}-{:#x}, is_managed", start, end);
-                } else {
-                    log::trace!("Unpoisoning: {:#x}-{:#x}", start, end);
-                    self.map_shadow_for_region(start, end, true);
+                if start >= self.shadow_offset {
+                    log::trace!("Reached the shadow/allocator region - stopping");
+                    return false;
```
Comment (s1341) [inline libafl_frida/src/alloc.rs:574]: I'm not 100% certain that there can be nothing allocated above the shadow/allocator region, under all circumstances. Can we not simply skip that region and continue afterwards if there is anything there? Seems safer?

File: libafl_frida/src/asan/asan_rt.rs [hunk 4]
```diff
@@ -514,18 +488,13 @@ impl AsanRuntime {
                         //is this necessary? The stalked return address will always be the real return address
                      //   let real_address = this.real_address_for_stalked(invocation.return_addr());
                         let original = [<$name:snake:upper _PTR>].get().unwrap();
-                        if this.hooks_enabled {
-                            let previous_hook_state = this.hooks_enabled;
-                            this.hooks_enabled = false;
+                        if !ASAN_IN_HOOK.get() && this.hooks_enabled { //not in hook and hooks are enabled
```
Comment (s1341) [inline libafl_frida/src/asan/asan_rt.rs:491]: Why did you change `this.hooks_enabled` to `ASAN_IN_HOOK.get()`?

File: libafl_frida/src/asan/asan_rt.rs [hunk 5]
```diff
@@ -600,8 +569,9 @@ impl AsanRuntime {
                         let mut invocation = Interceptor::current_invocation();
                         let this = &mut *(invocation.replacement_data().unwrap().0 as *mut AsanRuntime);
                         let original = [<$name:snake:upper _PTR>].get().unwrap();
-
-                        if !ASAN_IN_HOOK.get() && this.hooks_enabled && this.[<hook_check_ $name>]($($param),*){
+                        //don't check if hooks are enabled as there are certain cases where we want to run the hook even if we are out of the program
```
Comment (s1341) [inline libafl_frida/src/asan/asan_rt.rs:572]: Can you elaborate? I don't quite understand.

**Comment by domenukk** [comment]
@s1341 @mineo333 status? :)

**Comment by s1341** [comment]
@domenukk waiting for @mineo333 to respond to last review.

**Comment by mineo333** [comment]
Sorry, been fairly busy for the last few days. Gonna try to wrap it up today.

**Comment by s1341** [comment]
no worries.

File: libafl_frida/src/asan/asan_rt.rs (hunk 4 above)
Comment (mineo333) [reply libafl_frida/src/asan/asan_rt.rs]: There is an issue in the test with`operator new`. `operator new` runs with the second matching case for hook which sets `ASAN_IN_HOOK` to false. However, `hooks_enabled` is still true, so when mmap runs in `alloc`, it his `hook_mmap` which unpoisons the entire allocated region causing the test to fail.

File: libafl_frida/src/asan/asan_rt.rs (hunk 5 above)
Comment (mineo333) [reply libafl_frida/src/asan/asan_rt.rs:572]: The case that I was facing is that very rarely there would be a case where while hooks were enabled and the fuzzer code was running, the fuzzer would make an allocation. This allocation would be made in our allocator. However, this allocation would be freed when hooks were disabled. Thus, it would try to free our own allocation in the normal allocator which an issue.

**Comment by mineo333** [comment]
The new solution basically uses the darwin API directly instead of going through Frida. I'm unsure why Frida is so slow as this is a very similar solution to what Frida's enumerate_modules does. 

I suspect that it is because my implementation is not exploring submaps, but I'm not sure.

**Comment by domenukk** [comment]
@s1341 looks good?

File: libafl_frida/src/alloc.rs [hunk 6]
```diff
@@ -562,14 +578,70 @@ impl Allocator {
...
+            if kr != KERN_SUCCESS {
+                break;
+            }
+
+            let start = address as usize;
+            let end = (address + size) as usize;
+
+            if info.protection & VM_PROT_READ == VM_PROT_READ {
+                //if its at least readable
+                if self.shadow_offset <= start && end <= self.current_mapping_addr {
```
Comment (s1341) [inline libafl_frida/src/alloc.rs:625]: shouldn't the right hand side of the second condition be `self.base_mapping_addr`?

Comment (s1341) [reply libafl_frida/src/alloc.rs:625]: oh wait. no never mind.

File: libafl_frida/src/asan/asan_rt.rs (hunk 5 above)
Comment (s1341) [reply libafl_frida/src/asan/asan_rt.rs:572]: Ok. We will have to check this behavior on windows...

**Review by s1341 — APPROVED** [review]
Look's good. We can merge.
